lib/experimental/block-editor-settings.php

```diff
@@ -0,0 +1,87 @@
+<?php
+/**
+ * Adds settings to the block editor.
+ *
+ * @package gutenberg
+ */
+
+/**
+ * Adds styles and __experimentalFeatures to the block editor settings.
+ *
+ * @param array $settings Existing block editor settings.
+ *
+ * @return array New block editor settings.
+ */
+function gutenberg_get_block_editor_settings_experimental( $settings ) {
+	$is_block_theme = wp_is_block_theme();
+
+	global $post_id;
+
+	if ( ! $is_block_theme || ! $post_id ) {
+		return $settings;
+	}
+
+	$template_slug = get_page_template_slug( $post_id );
+
+	if ( ! $template_slug ) {
+		$post_slug      = 'singular';
+		$page_slug      = 'singular';
+		$template_types = get_block_templates();
+
+		foreach ( $template_types as $template_type ) {
+			if ( 'page' === $template_type->slug ) {
+				$page_slug = 'page';
+			}
+			if ( 'single' === $template_type->slug ) {
+				$post_slug = 'single';
+			}
+		}
+
+		$what_post_type = get_post_type( $post_id );
```

Comment:
I think this is _nearly_ there — it turns out the global variable for the post id set by `post-new.php` is `$post_ID` (capitalised ID) on this line: https://github.com/WordPress/wordpress-develop/blob/6742d0d7a65e37f24c67eeaae373f4c086c277a5/src/wp-admin/post-new.php#L67

I think we'll need to use `global $post_ID` instead, otherwise `$post_id` isn't set when loading the page for creating a new post. For a new post we wind up with a layout shift during loading, once the editor has had a chance to resolve the template:

![2023-03-07 14 38 14](https://user-images.githubusercontent.com/14988353/223314901-7dc9fad6-7c01-42cd-b20d-2d19ec28b6d4.gif)

But it looks to work pretty well locally if I swap it out for `$post_ID` instead 🙂

Reply:
Oh that's weird, but it does seem to work! Updating.